### ISSUE TITLE
Thoughts on code-signing from wjk

### ISSUE THREAD

**Issue by wjk** [octave-app/octave-app#196]
While investigating macOS code-signing, I discovered that apps that contain relative symbolic links that point inside the bundle _are_ now honored by the code signing system. To test this, 
I downloaded the (unsigned) Octave-5.2.0.app, made a scratch copy, and signed it using my Apple Development certificate. It signed correctly. The output of `codesign -d -vvvv` is included at the end of this post. I am running macOS 10.15.4 and Xcode 11.4.1.

---
```
Executable=/Users/wjk/Documents/Octave-5.2.0.app/Contents/MacOS/applet
Identifier=org.octave-app.Octave
Format=app bundle with Mach-O universal (i386 x86_64)
CodeDirectory v=20200 size=277 flags=0x0(none) hashes=3+3 location=embedded
VersionPlatform=1
VersionMin=656896
VersionSDK=658688
Hash type=sha256 size=32
CandidateCDHash sha1=3e895210975eca90a881f0e7f1856f9d58043b45
CandidateCDHashFull sha1=3e895210975eca90a881f0e7f1856f9d58043b45
CandidateCDHash sha256=00223be9dc31f22e245a6eac23958e31d1dc1733
CandidateCDHashFull sha256=00223be9dc31f22e245a6eac23958e31d1dc17330af3f096e4412a3934343655
Hash choices=sha1,sha256
CMSDigest=b3c5be38e9f1e9472be1f23b716d1149fbe654a04fa4b4a96bbe6d0b38abceb8
CMSDigestType=2
Page size=4096
CDHash=00223be9dc31f22e245a6eac23958e31d1dc1733
Signature size=4817
Authority=Apple Development: William Kent (47T5C6TZYT)
Authority=Apple Worldwide Developer Relations Certification Authority
Authority=Apple Root CA
Signed Time=May 1, 2020 at 5:21:37 PM
Info.plist entries=16
TeamIdentifier=3P242C9ES5
Sealed Resources version=2 rules=13 files=71447
Internal requirements count=1 size=184
```

**Comment by apjanke** [comment]
Sweet! I'll see if this works on 10.13 now, too. (We build on 10.13 so we can support multiple user OS versions.) Thanks for the tip!

**Comment by apjanke** [comment]
Oh, BTW, I saw your idea about dlopen()/dlsym()/trampolining into the Octave GUI instead of running it as a child process in the original GitHub notification email for this bug, and I've had some time to digest it. That sounds really interesting – I think the multi-process model is causing some UI hinkiness on Mac, with respect to multiple/wrong Dock icons or lingering processes – but I think modifying the actual built Octave binary that much is beyond the scope of the packaging work that Octave.app does. (And it's over my head so I can't really evaluate it.) But that's interesting enough that it might be worth pitching directly to the upstream Octave developers. I'd encourage you to send it to the [Octave Maintainers mailing list](https://lists.gnu.org/mailman/listinfo/octave-maintainers) if it's something you're interested in pursuing.

**Comment by wjk** [comment]
@apjanke Thanks for the good word. However, sending patches to Octave itself is not something I really am interested in pursuing. The idea came to me while I was working on the code signing issue, but it wasn’t something I really wanted to invest that much time in honestly.

While my POC implementation did not require modifying Octave itself, I did have to directly reference (hardcode the bundle-relative path to) the octave-gui binary under the libexec directory, which I believe is a private implementation detail of Octave. When I tried dlopen()ing the “public” octave binary in the bin directory, it did me no good, as that binary just forks/execs the octave-gui file, resulting in two Dock icons instead of the one I was going for. Furthermore, I didn’t think this would be an acceptable technique for the released version of the app, as it would mean the launcher program in the MacOS directory would need to be updated every time Octave is, since the path to octave-gui contains the version number.
